src/main/java/com/hivemq/extensions/handler/PubackInterceptorHandler.java

```diff
@@ -0,0 +1,441 @@
+                continue;
+            }
+            final PubackInboundInterceptorTask interceptorTask =
+                    new PubackInboundInterceptorTask(interceptor, interceptorFuture, plugin.getId());
+
+            pluginTaskExecutorService.handlePluginInOutTaskExecution(
+                    interceptorContext, input, output, interceptorTask);
+        }
+
+        final InboundInterceptorFutureCallback callback =
+                new InboundInterceptorFutureCallback(output, puback, ctx);
+        Futures.addCallback(interceptorFuture, callback, ctx.executor());
+        return true;
+    }
+
+    private static class InboundInterceptorFutureCallback implements FutureCallback<Void> {
+
+        private final @NotNull PubackInboundOutputImpl output;
+        private final @NotNull PUBACK puback;
+        private final @NotNull ChannelHandlerContext ctx;
+
+        InboundInterceptorFutureCallback(
+                final @NotNull PubackInboundOutputImpl output,
+                final @NotNull PUBACK puback,
+                final @NotNull ChannelHandlerContext ctx) {
+            this.output = output;
+            this.puback = puback;
+            this.ctx = ctx;
+        }
+
+        @Override
+        public void onSuccess(final @Nullable Void result) {
+            try {
+                final PUBACK finalPuback = PUBACK.createPubackFrom(output.getPubackPacket());
+                ctx.fireChannelRead(finalPuback);
+            } catch (final Exception e) {
+                log.error("Exception while modifying an intercepted PUBACK message.", e);
+                ctx.fireChannelRead(puback);
+            }
+        }
+
+        @Override
+        public void onFailure(final @NotNull Throwable t) {
+            //should never happen, since the settable future never sets an exception
+            ctx.channel().close();
+        }
+    }
+
+    private static class PubackInboundInterceptorContext extends PluginInOutTaskContext<PubackInboundOutputImpl> {
+
+        private final @NotNull PubackInboundInputImpl input;
+        private final @NotNull PubackInboundOutputImpl output;
+        private final @NotNull SettableFuture<Void> interceptorFuture;
+        private final int interceptorCount;
+        private final @NotNull AtomicInteger counter;
+
+        PubackInboundInterceptorContext(
+                final @NotNull Class<?> taskClazz,
+                final @NotNull String clientId,
+                final @NotNull PubackInboundInputImpl input,
+                final @NotNull PubackInboundOutputImpl output,
+                final @NotNull SettableFuture<Void> interceptorFuture,
+                final int interceptorCount) {
+            super(taskClazz, clientId);
+            this.input = input;
+            this.output = output;
+            this.interceptorFuture = interceptorFuture;
+            this.interceptorCount = interceptorCount;
+            this.counter = new AtomicInteger(0);
```

Comment:
Why is this an atomic integer?
For a single client this should not be called concurrently